File: 0821. Shortest Distance to a Character/0821.py
# time complexity: O(n) - two linear scans over the string
# space complexity: O(n)
class Solution(object):
    # two-pass
    def shortestToChar(self, s, c):
        """
        :type s: str
        :type c: str
        :rtype: List[int]
        """
        n = len(s)
        res = [0] * n
        prev = -n                    # position of last seen c (far left)
        
        # left -> right pass: distance to previous c
        for i in range(n):
            if s[i] == c:
                prev = i             # update last seen c
            res[i] = i - prev        # distance from i to last c on left
        
        prev = 2 * n                 # position of next c (far right)
        # right -> left pass: take min with next c on right
        for i in range(n-1, -1, -1):
            if s[i] == c:
                prev = i             # update next seen c
            res[i] = min(res[i], prev - i)
        
        return res

File: 0821. Shortest Distance to a Character/0821.java
// time complexity: O(n)
// space complexity: O(n)
class Solution {
    // two-pass scan
    public int[] shortestToChar(String s, char c) {
        int n = s.length();
        int[] answer = new int[n];
        int prev = -n;                // position of last seen c (far left initially)

        // left to right pass: distance to previous c
        for (int i=0; i<n; i++) {
            if (s.charAt(i) == c) {
                prev = i;             // update last seen c
            }
            answer[i] = i - prev;        // distance from i to last c on left
        }

        prev = 2*n;                 // position of next c (far right initially)
        // right to left pass: take min with next c on right
        for (int i=n-1; i>=0; i--) {
            if (s.charAt(i) == c) {
                prev = i;             // update next seen c
            }
            answer[i] = Math.min(answer[i], prev - i);
        }

        return answer;
    }
}
